pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,859 @@
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture. In the PacketCapture spec, at least one of `.Spec.Source.Pod` or `.Spec.Destination.Pod`
+// should be set.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) error {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+	var err error
+	pcState := &packetCaptureState{name: name}
+	cleanup := func() {
+		if err != nil {
+			klog.ErrorS(err, "PackageCapture failed", "name", name)
+			err := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
+			if err != nil {
+				klog.ErrorS(err, "Failed to update PacketCapture status")
+			}
+		}
+	}
+	defer cleanup()
+	pc, err := c.packetCaptureLister.Get(name)
+	if err != nil {
+		return err
+	}
+	device := c.getTargetCaptureDevice(pc)
+	klog.V(2).InfoS("Prepare capture on current node", "name", pc.Name, "device", *device)
+	packetSource, err := c.newPacketSourceFn(c, pc)
+	if err != nil {
+		klog.ErrorS(err, "Failed to create packet source")
+		return err
+	}
+	pcState.targetCapturedPacketsNum = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	timeout := defaultTimeoutDuration
+	if pc.Spec.Timeout != nil {
+		timeout = time.Duration(*pc.Spec.Timeout) * time.Second
+	}
+	go func() error {
+		err = c.performCapture(pcState, timeout, packetSource)
+		func() {
+			c.cond.L.Lock()
+			defer c.cond.L.Unlock()
+			c.captures[name] = packetCapturePhaseCompleted
+			c.numRunningCaptures -= 1
+			c.cond.Signal()
+		}()
+
+		updateErr := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
+		if updateErr != nil {
+			klog.ErrorS(updateErr, "Failed to update PacketCapture status")
+		}
+		return err
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(captureState *packetCaptureState, timeout time.Duration, source PacketSource) error {
+	options := CaptureOptions{
+		MaxCaptureLength: maxSnapshotBytes,
+		Promiscuous:      false,
+	}
+	packets, err := source.Capture(&options)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+
+	timer := time.NewTicker(timeout)
+	defer timer.Stop()
```

Comment:
It should create a context with the timeout and pass to `Capture`, otherwise the spawned goroutine would be leaked.